### PR TITLE
update pipeline labels for github issues

### PR DESCRIPTION
**Pull request by heng-liu**
Fix https://github.com/NuGet/Home/issues/9133
Update pipeline labels for GitHub issues.

### CODE REVIEW

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        private string githubToken;
+        private string zenhubToken;
+        private int issueNumberFrom;
+        private int issueNumberTo;
+
+        //If specify a range in the command, then only run issues having issue numbers in this range.
+        //If range is not specified, run all the issues in the repository.
+        bool processAll;
+
+        //private SummaryReport summaryReport;
```
Comment (zkat) [inline UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:32]: leftover comment?

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    {
+        private const string WorkspaceName = "NuGet Client Team";
+        private readonly GitHubClient GitHubClient;
+        private readonly ZenHubClient ZenHubClient;
+
+        private string repoName;
+        private string githubToken;
+        private string zenhubToken;
+        private int issueNumberFrom;
+        private int issueNumberTo;
```
Comment (zkat) [inline UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:26]: shouldn't properties be PascalCase?

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+            string[] repoParts = repoName.Split('/');
+
+            List<string> pipelineLabels = new List<string>();
+            pipelineLabels.Add(IssueLabels.NewIssues);
+            pipelineLabels.Add(IssueLabels.Icebox);
+            pipelineLabels.Add(IssueLabels.Backlog);
+            pipelineLabels.Add(IssueLabels.InProgress);
+            pipelineLabels.Add(IssueLabels.InReview);
+            pipelineLabels.Add(IssueLabels.Validating);
```
Comment (zkat) [inline UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:70]: Can't you build Lists like this?

```suggestion
            List<string> pipelineLabels = new List<string>()
            {
                IssueLabels.NewIssues,
                IssueLabels.Icebox,
                ...etc
            };
```

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+                    try
+                    {
+                        await GitHubClient.Issue.Update(repoParts[0], repoParts[1], issue.Number, issueUpdate);
+                        Console.WriteLine($"Updated : issue {issue.HtmlUrl} ");
+                        updatedCount++;
+
+                        //According to https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits
+                        //If you're making a large number of POST, PATCH, PUT, or DELETE requests for a single user or client ID, wait at least one second between each request.
+                        System.Threading.Thread.Sleep(1000);
```
Comment (zkat) [inline UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:142]: I think you can safely remove this -- the authenticated API limit is 5k requests per hour. We could run this script multiple times an hour and still not hit that cap: https://developer.github.com/v3/#rate-limiting

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 1 above)
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:32]: Thanks for reviewing! Yes, should be removed.

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 2 above)
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:26]: Thanks! Fixed.

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 3 above)
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:70]: Thanks! Changed.

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 4 above)
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:142]: Thanks, but I'm not sure if this is an additional limitation for PATCH request: (updating label is a PATCH request)
https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits
`If you're making a large number of POST, PATCH, PUT, or DELETE requests for a single user or client ID, wait at least one second between each request.`

File: UpdatePipelineLabel/UpdatePipelineLabel/IssueLables.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+        public static string Icebox = "Pipeline: Icebox";
+
+        public static string Backlog = "Pipeline: Backlog";
+
+        public static string InProgress = "Pipeline: In Progress";
+
+        public static string InReview = "Pipeline: In Review";
+
+        public static string Validating = "Pipeline: Validating";
```
[the remark this answers is not in the document]
Comment (nkolev92) [reply UpdatePipelineLabel/UpdatePipelineLabel/IssueLables.cs:19]: Yep makes sense.

**Comment by nkolev92** [comment]
@heng-liu 
I imagine this is something we run infrequently? Is it automated or run manually?

File: UpdatePipelineLabel/UpdatePipelineLabel/IssueLables.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/IssueLables.cs:19]: Thanks! Updated.

**Comment by heng-liu** [comment]
@nkolev92 , it is not automatied, we need to manually run the command, to update the pipeline labels for issues.

**Comment by heng-liu** [comment]
It will only run on the issues not with the right pipeline label. Issues with correct pipeline labels will be skipped. So it's like running incrementally.

**Comment by nkolev92** [comment]
> @nkolev92 , it is not automatied, we need to manually run the command, to update the pipeline labels for issues.


Who's gonna run this? You or is it the backlog gardeners?

**Comment by heng-liu** [comment]
Sorry, I'm not sure what is the backlog gardeners.. Shall I set up something to make it be able to run by bots or something?

**Comment by nkolev92** [comment]
> Sorry, I'm not sure what is the backlog gardeners

Look for NuGet Client Backlog Gardeners in your e-mails :) 

Anyway, if it's easy, do it, but I think manual is fine for now, especially given that if you set-up automation you have to protect the key etc, etc.

**Comment by nkolev92** [comment]
My bad, miss-click :D

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 4 above)
Comment (zkat) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs:148]: Are you running into actual errors, then?

**Comment by zkat** [comment]
@nkolev92 it's probably out of scope for this particular work, but we can (eventually?) set up a service that runs this per-issue whenever an issue is transferred between pipelines: https://github.com/ZenHubIO/API#issue-transfer

**Comment by nkolev92** [comment]
@zkat Yeah that's create.
Let's track that in the CE repo.

**Comment by zkat** [comment]
@heng-liu can you create an issue to track setting up a service? :)

**Comment by heng-liu** [comment]
@zkat Sure! Hi, @nkolev92 the setting up a service issue should be in CE repo, right?

**Comment by nkolev92** [comment]
@heng-liu yes.

File: UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs (hunk 4 above)
Comment (heng-liu) [reply UpdatePipelineLabel/UpdatePipelineLabel/LabelUpdateHelper.cs]: Thanks Kat! I tested with 100+ updates without sleep, and it succeeded. Removed sleep.

**Pull request merged**